### PR TITLE
Build Docker images on multiple cuda versions in parallel

### PR DESCRIPTION
**Pull request by DerrickYLJ**
**Description of changes:**
Parallelize the process of building docker images with multiple different cuda versions as base images and modify the build.sh, pulish.sh, and workflow to distinguish the tag of cuda or hip_rocm as FF_BACK_END.

**Related Issues:**

Linked Issues:
- Issue #

Issues closed by this PR:
- Closes #

**Before merging:**

- [ ] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules?

### CODE REVIEW

File: docker/run.sh [hunk 1]
```diff
@@ -15,6 +15,15 @@ SHM_SIZE=8192m
 gpu_arg=""
 if $ATTACH_GPUS ; then gpu_arg="--gpus all" ; fi
 image=${1:-flexflow}
+cuda_version=${cuda_version:-11.1.1}
+
+if [[ "$cuda_version" != @(11.1.1|11.2.2|11.3.1|11.5.2|11.6.2|11.7.1|11.8.0) ]]; then
```
Comment (goliaro) [inline docker/run.sh:20]: could we also autodetect the cuda version in the `run.sh`? And could you add a message telling the user which cuda version the script is building for ? E.g. "Building FlexFlow docker image for CUDA 11.8"

File: docker/build.sh [hunk 2]
```diff
@@ -8,54 +8,16 @@ SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
 # Cd into $FF_HOME. Assumes this script is in $FF_HOME/docker
 cd "$SCRIPT_DIR/.."
 
-cuda_version="empty"
-image="flexflow"
+cuda_version=${cuda_version:-11.1.1}
```
Comment (goliaro) [inline docker/build.sh:11]: can we make the `cuda_version` environment variable take CUDA versions with only one decimal, like the previous cuda_version argument? then, we can add the default subversion digit in the script

File: docker/build.sh [hunk 3]
```diff
@@ -8,54 +8,16 @@ SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
...
   # validate the verison of CUDA against a list of supported ones
   # 11.1, 11.3, 11.5, 11.6, 11.7, 11.8
-  # Available versions: 11.1.1 | 11.2.2 | 11.3.1 | 11.5.2 | 11.6.2 | 11.7.1 | 11.8.0
-  echo "cuda_version is not supported, please choose among {11.1,11.3,11.5,11.6,11.7,11.8}"
+  echo "cuda_version is not supported, please choose among {11.1.1|11.2.2|11.3.1|11.5.2|11.6.2|11.7.1|11.8.0}"
   exit 1
 fi
 
-# modify cuda version to available versions
-if [[ "$cuda_version" == @(11.1|11.3|11.7) ]]; then
```
Comment (goliaro) [inline docker/build.sh:50]: I think we can bring back this block

**Review by goliaro — CHANGES_REQUESTED** [review]

File: docker/run.sh [hunk 4]
```diff
@@ -27,16 +36,32 @@ else
...
-    eval docker run -it "$gpu_arg" "--shm-size=${SHM_SIZE}" "flexflow-environment-${FF_GPU_BACKEND}:latest"
-elif [[ "$image" == "flexflow" ]]; then
-    eval docker run -it "$gpu_arg" "--shm-size=${SHM_SIZE}" "flexflow-${FF_GPU_BACKEND}:latest"
-elif [[ "$image" == "mt5" ]]; then
-    # Backward compatibility
-    eval docker run -it "$gpu_arg" "--shm-size=${SHM_SIZE}" \
-    -v "$(pwd)"/../examples/python/pytorch/mt5/data:/usr/FlexFlow/examples/python/pytorch/mt5/data \
-    -v "$(pwd)"/../examples/python/pytorch/mt5/eng-sin.tar:/usr/FlexFlow/examples/python/pytorch/mt5/eng-sin.tar \
-    "flexflow-${FF_GPU_BACKEND}:latest"
+if [[ "${FF_GPU_BACKEND}" == "hip_rocm" ]]; then
```
Comment (goliaro) [inline docker/run.sh:39]: I think here, to reduce the code duplication, we could use the `if [[ "${FF_GPU_BACKEND}" == "hip_rocm" ]]; then ... else ... fi` statement only to modify the `cuda_version` variable. For rocm, we can make it an empty string, whereas for CUDA, we can set it as `cuda_version="-${cuda_version}"` (we add the hyphen). Then, we can run the docker machines "flexflow-environment-${FF_GPU_BACKEND}${cuda_version}:latest", "flexflow-${FF_GPU_BACKEND}${cuda_version}:latest", etc... in all cases. Note how I have removed the hyphen between `${FF_GPU_BACKEND}` and `${cuda_version}`

**Comment by DerrickYLJ** [comment]
Most changes have been completed except for the "autodetect cuda_version in run.sh". I added back the `--cuda_version` optional flag so that users are able to input any cuda versions they want to build, and as before, if there is no input, it will autodetect the cuda version installed on the current machine and run with that if available.

File: docker/README.md [hunk 5]
```diff
@@ -35,16 +35,16 @@ FF_GPU_BACKEND=<YOUR_GPU_BACKEND> ./docker/build.sh flexflow-environment
 ``` 
 
 ## Running a Docker container
-After having either built or downloaded a Docker container by following the instructions above, you can run it with the following command (the `flexflow` argument of the run script can be omitted):
+After having either built or downloaded a Docker container by following the instructions above, you can run it with the following command (the `--image_name flexflow` argument of the run script can be omitted). Moreover, you can choose the CUDA version to build using optional flag `--cuda_version`; otherwise, it will autodetect the cuda_version installed on the current machine and run the Docker container with it if available.
```
Comment (goliaro) [inline docker/README.md:38]: can we update the readme to reflect the new way of passing the CUDA version, and image name?

File: docker/README.md [hunk 6]
```diff
@@ -35,16 +35,16 @@ FF_GPU_BACKEND=<YOUR_GPU_BACKEND> ./docker/build.sh flexflow-environment
...
 
 ```
-FF_GPU_BACKEND=<YOUR_GPU_BACKEND> ./docker/run.sh flexflow 
+FF_GPU_BACKEND=<YOUR_GPU_BACKEND> ./docker/run.sh --image_name flexflow --cuda_version 11.1
```
Comment (goliaro) [inline docker/README.md:41]: same here

File: docker/build.sh [hunk 7]
```diff
@@ -8,55 +8,24 @@ SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
...
-      ;;
-    *)
-      echo "Invalid option: $key"
-      exit 1
-      ;;
-  esac
-done
-
-if [[ $cuda_version == "empty" ]]; then
-  cuda_version=$(command -v nvcc >/dev/null 2>&1 && nvcc --version | grep "release" | awk '{print $NF}')
```
Comment (goliaro) [inline docker/build.sh:35]: I think we should bring back this block to perform autodetection

File: docker/build.sh [hunk 8]
```diff
@@ -74,7 +43,12 @@ else
 fi
 
 # Build the FlexFlow Enviroment docker image with input cuda version
-docker build --build-arg "FF_GPU_BACKEND=${FF_GPU_BACKEND}" --build-arg "cuda_version=${cuda_version}" -t "flexflow-environment-${FF_GPU_BACKEND}-${cuda_version}" -f docker/flexflow-environment/Dockerfile .
+# if FF_GPU_BACKEND, we don't tag the docker image with cuda version 
+if [[ "${FF_GPU_BACKEND}" == "hip_rocm" ]]; then
```
Comment (goliaro) [inline docker/build.sh:47]: here, to reduce code duplication, we can use the `if [[ "${FF_GPU_BACKEND}" == "hip_rocm" ]];` to set the cuda_version to variable to the empty string, and then use the same docker build command for both hip_rocm and CUDA

File: docker/build.sh [hunk 9]
```diff
@@ -132,4 +106,8 @@ fi
 # Set value of BUILD_CONFIGS
 get_build_configs
 
-docker build --build-arg "N_BUILD_CORES=${n_build_cores}" --build-arg "FF_GPU_BACKEND=${FF_GPU_BACKEND}" --build-arg "BUILD_CONFIGS=${BUILD_CONFIGS}" --build-arg "cuda_version=${cuda_version}" -t "flexflow-${FF_GPU_BACKEND}-${cuda_version}" -f docker/flexflow/Dockerfile .
\ No newline at end of file
+if [[ "${FF_GPU_BACKEND}" == "hip_rocm" ]]; then
+  docker build --build-arg "N_BUILD_CORES=${n_build_cores}" --build-arg "FF_GPU_BACKEND=${FF_GPU_BACKEND}" --build-arg "BUILD_CONFIGS=${BUILD_CONFIGS}" -t "flexflow-${FF_GPU_BACKEND}" -f docker/flexflow/Dockerfile .
```
Comment (goliaro) [inline docker/build.sh:110]: same here. After having set the cuda_version to the empty string above when building for hip_rocm, here we can remove this if statement and use a single docker build command for both

File: docker/flexflow/Dockerfile [hunk 10]
```diff
@@ -1,6 +1,6 @@
 ARG FF_GPU_BACKEND "cuda"
-ARG cuda_version
-FROM flexflow-environment-$FF_GPU_BACKEND-$cuda_version:latest
+ARG cuda_version ""
+FROM flexflow-environment-$FF_GPU_BACKEND$cuda_version:latest
```
Comment (goliaro) [inline docker/flexflow/Dockerfile:3]: perfect!

File: docker/publish.sh [hunk 11]
```diff
@@ -4,59 +4,28 @@ set -euo pipefail
...
-      shift 2
-      ;;
-    *)
-      echo "Invalid option: $key"
-      exit 1
-      ;;
-  esac
-done
-
-if [[ $cuda_version == "empty" ]]; then
```
Comment (goliaro) [inline docker/publish.sh:30]: we should bring back this block for automatic detection

File: docker/run.sh [hunk 12]
```diff
@@ -27,16 +61,23 @@ else
...
 
+if [[ "${FF_GPU_BACKEND}" == "hip_rocm" ]]; then
+  cuda_version=""
+else
+  cuda_version="-${cuda_version}"
+fi
+  
 if [[ "$image" == "flexflow-environment" ]]; then
-    eval docker run -it "$gpu_arg" "--shm-size=${SHM_SIZE}" "flexflow-environment-${FF_GPU_BACKEND}:latest"
+    eval docker run -it "$gpu_arg" "--shm-size=${SHM_SIZE}" "flexflow-environment-${FF_GPU_BACKEND}${cuda_version}:latest"
```
Comment (goliaro) [inline docker/run.sh:71]: great! that's exactly what I had in mind here

**Review by goliaro — CHANGES_REQUESTED** [review]

**Review by goliaro — APPROVED** [review]
I just made some small additional fixes / reorderings. LGTM!

**Pull request merged**